frame/merkle-mountain-range/src/lib.rs

```diff
@@ -318,38 +318,75 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
 			.saturating_add(leaf_index.saturated_into())
 	}
 
-	/// Generate a MMR proof for the given `leaf_indices`.
+	/// Convert a `block_num` into a leaf index.
+	fn block_num_to_leaf_index(block_num: T::BlockNumber) -> Result<LeafIndex, primitives::Error>
+	where
+		T: frame_system::Config,
+	{
+		// leaf_indx = block_num - (current_block_num - leaves_count) - 1;
+		let leaves_count = Self::mmr_leaves().saturated_into::<u64>();
+		let current_block_num = <frame_system::Pallet<T>>::block_number().saturated_into::<u64>();
+		let diff = current_block_num.saturating_sub(leaves_count.into());
+
+		let block_num_as_u64 = block_num.saturated_into::<u64>();
+
+		if block_num_as_u64 <= diff {
+			return Err(
+				primitives::Error::BlockNumToLeafIndex.log_debug("The block_number is incorrect.")
+			)
+		}
+
+		let leaf_index = (block_num_as_u64.saturating_sub(diff).saturating_sub(1u32.into()))
+			.saturated_into::<LeafIndex>();
+		Ok(leaf_index)
+	}
```

Comment:
I would:
- minimize the number of conversions between types
- convert from one type to the other with `try_into()` instead of saturated_into()

```
	fn block_num_to_leaf_index(block_num: T::BlockNumber) -> Result<LeafIndex, primitives::Error>
	where
		T: frame_system::Config,
		T::BlockNumber: TryInto<LeafIndex>,
	{
		// leaf_idx = (leaves_count - 1) - (current_block_num - block_num);
		let best_block_num = <frame_system::Pallet<T>>::block_number();
		let blocks_diff = best_block_num.checked_sub(&block_num).ok_or_else(|| {
			primitives::Error::BlockNumToLeafIndex.log_debug(
				"The provided block_number is greater than the best block number.",
			)
		})?;
		let blocks_diff_as_leaf_idx = blocks_diff.try_into().map_err(|_| {
			primitives::Error::BlockNumToLeafIndex
				.log_debug("The `blocks_diff` couldn't be converted to `LeafIndex`.")
		})?;

		let leaf_idx = Self::mmr_leaves()
			.checked_sub(1)
			.and_then(|last_leaf_idx| last_leaf_idx.checked_sub(blocks_diff_as_leaf_idx))
			.ok_or_else(|| {
				primitives::Error::BlockNumToLeafIndex
					.log_debug("There aren't enough leaves in the chain.")
			})?;
		Ok(leaf_idx)
	}
```

Not sure if we need to do `checked_sub` for the `leaf_idx`. `saturating_sub` might be enough. I didn't think a lot about this aspect.

Reply:
> Actually there are 2 edge cases:
> 
> 1. The provided block number is invalid (too big) -> this is covered
> 2. The provided block number is valid but it's too smal (at that block, the mmr pallet wasn't activated yet)

case `2.` is handled in the current version of `block_num_to_leaf_index` by this:
https://github.com/paritytech/substrate/blob/4538c0832e81d1bff5df9ff821c5515326bcc54e/frame/merkle-mountain-range/src/lib.rs#L333-L337
and imo if you're explicitly converting a block number to a leaf index, this should error if there can't be a leaf for said block. or we convert it to an option, like @serban300 proposes. that's a good option too - we just lose some handling granularity that having distinct errors for `1.` and `2.` would provide (if `1.` would return an option too).

for case `1.`, my weakly held opinion is that this should not be checked inside `block_num_to_leaf_index`, but handled at the call site, like it is currently
https://github.com/paritytech/substrate/blob/4538c0832e81d1bff5df9ff821c5515326bcc54e/frame/merkle-mountain-range/src/lib.rs#L375-L379
since the leaf index for a provided block number can be valid, it's just that the block may not exist yet, or the runtime is not aware of it yet, and this can be a valid use case (it's just that producing historic batch proofs is not, which is why that specific use case should error).

Reply:
> for case 1., my weakly held opinion is that this should not be checked inside block_num_to_leaf_index, but handled at the call site, like it is currently

This might work too. I incline towards checking this in `block_num_to_leaf_index()`, but I don't have strong preferences.